rules/recording.go

```diff
@@ -93,6 +93,16 @@ func (rule *RecordingRule) Eval(ctx context.Context, ts time.Time, query QueryFu
 
 		sample.Metric = lb.Labels()
 	}
+
+	// Check that the rule does not produce identical metrics after applying
+	// labels #5529
+	if vector.ContainsSameLabelset() {
+		err = fmt.Errorf("resulting vector cannot contain metrics with the same labelset")
```

Comment:
I'd make this more in line with the error inside PromQL, but also indicating that this is related to templating and not PromQL per-se.